### PR TITLE
connlib: Improve FFI bridges for Apple and Android

### PR DESCRIPTION
**Pull request by francesca64**
This makes it possible to build the Apple/Android FFI bridges and integrate them with their respective client apps.

### CODE REVIEW

**Comment by jamilbk** [comment]
Adding @thomaseizinger for a quick scan for context as well

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (thomaseizinger) [inline rust/connlib/libs/tunnel/src/tun_android.rs:44]: It is 100% preference so feel free to discard but I find that using early return and avoiding else to be much easier to read:

```suggestion
        if fd == -1 {
            return Err(Error::Io(std::io::Error::last_os_error()))
        }
        
        Ok(Self { fd })
```

File: rust/connlib/libs/common/src/session.rs
Comment (thomaseizinger) [inline rust/connlib/libs/common/src/session.rs]: Having trait bounds on structs can become really cumbersome because they are quite infectious to the rest of the codebase. Unless you are referencing associated types (`Callbacks` doesn't have any from what I can see?) somewhere, I wouldn't do it.

Traits are about behaviour and behaviour is specified on `impl` blocks. `struct` declarations should be for layout only.

It is weird initially to have a type-parameter to be "too open" for what you want users to use it but in reality, they can't do anything with a "wrongly" instantiated type anyway if you have the correct bounds on your `impl` blocks which is where you need them anyway to actually implement your stuff and call the trait functions.

File: rust/connlib/libs/common/src/session.rs
Comment (thomaseizinger) [inline rust/connlib/libs/common/src/session.rs:56]: This forces interior-mutablity on all implementors. It might be more ergonomic to either have connlib have a single owner of this or make the trait `&mut self` and `Arc<Mutex>` it inside connlib.

File: rust/connlib/clients/android/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/android/src/lib.rs]: +1 for referencing `trace!` etc with `log::`. I much prefer that too because it avoids import lines and actually reads quite nice.

File: rust/connlib/clients/apple/Sources/Connlib/CallbackHandler.swift
Comment (thomaseizinger) [inline rust/connlib/clients/apple/Sources/Connlib/CallbackHandler.swift:103]: Resolved now?

**Review by thomaseizinger — DISMISSED** [review]
Super clean commit separation, made it very easy to review this!

I can't comment on any of the ffi stuff :)

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [inline rust/connlib/clients/android/src/lib.rs]: We [may have renamed](https://github.com/firezone/product/blob/main/architecture/madr.md#2023_06_19-grooming) this callback to `onConnect` in the last grooming. Will defer to you though -- I don't feel strongly Re: naming conventions

File: rust/connlib/clients/apple/Sources/Connlib/CallbackHandler.swift
Comment (francesca64) [reply rust/connlib/clients/apple/Sources/Connlib/CallbackHandler.swift:103]: Errors need to be propagated to the client; the logging is only helpful for testing.

**Review by jamilbk — COMMENTED** [review]
Just the note regarding `onConnect`. Otherwise LGTM, looks a lot cleaner from what I can tell!

File: rust/connlib/libs/common/src/session.rs
Comment (francesca64) [reply rust/connlib/libs/common/src/session.rs:56]: We decided that the burden of thread-safety/etc. is on the non-Rust side: https://github.com/firezone/product/blob/main/architecture/madr.md#2023_06_14-client-app-call

On the Rust side, the underlying Apple/Android callback handler is just an opaque pointer to some Swift/Kotlin object.

File: rust/connlib/libs/tunnel/src/tun_android.rs
Comment (francesca64) [reply rust/connlib/libs/tunnel/src/tun_android.rs:44]: I empathize with your viewpoint, but I consider early return to be a larger hammer, and branching has the additional benefit of making it easier to visualize potential combinator golfs/etc.

File: rust/connlib/libs/common/src/session.rs
Comment (francesca64) [reply rust/connlib/libs/common/src/session.rs]: I understand your point, but I don't see the merit in `Session` having more degrees of freedom than we actually want or use. The type definition reads loud and clear: a `Session` where `CB` does not implement `Callbacks` is not a `Session`.

The fact that Rust requires trait bounds to be repeated does not change the fact that this type definition is my explicit intention. While your proposal may reduce visual noise, it increases conceptual/mental-model noise by introducing a degenerate case for `Session` types.

File: rust/connlib/clients/android/src/lib.rs
Comment (francesca64) [reply rust/connlib/clients/android/src/lib.rs]: Absolutely; I was trying to avoid adding something else onto an already large-ish PR, but that's a small enough addition that it'd be worth me adding now just to avoid additional churn for the clients.

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Ah gotcha!

File: rust/connlib/libs/common/src/session.rs
Comment (thomaseizinger) [reply rust/connlib/libs/common/src/session.rs]: > I understand your point, but I don't see the merit in `Session` having more degrees of freedom than we actually want or use.

Yeah, whether or not repeating the trait bounds is annoying depends a lot on how many layers of nesting you've got.

If your trait is object-safe, another way would be to use a trait-object instead of a type-parameter. That is equally expressive from a readability PoV but results in simpler type definitions.

Comment (conectado) [reply rust/connlib/libs/common/src/session.rs]: I'd rather not introduce trait-objects if we don't need to, that  would indicate some kind of runtime polymorphism that we don't actually need.

Comment (conectado) [reply rust/connlib/libs/common/src/session.rs]: > The fact that Rust requires trait bounds to be repeated does not change the fact that this type definition is my explicit intention. While your proposal may reduce visual noise, it increases conceptual/mental-model noise by introducing a degenerate case for Session types.

I agree with this but then we might want to remove the `'static` bound? That's just an implementation detail that we might even want to change in the future.

**Review by conectado — COMMENTED** [review]
looks great! will approve once the CI is passing.

**Comment by roop** [comment]
I reviewed this from the Apple FFI point of view. This looks great.

  - Don't we need an `on_disconnect()` / `onDisconnect()` in the FFI CallbackHandler?
  - We need a logging callback as well, but I'm guessing that might come later, and might be out of scope for this PR.

File: rust/connlib/clients/apple/build-xcframework-dev.sh
Comment (roop) [inline rust/connlib/clients/apple/build-xcframework-dev.sh]: Maybe we should make this build for both macOS and iOS by default?
```suggestion
for sdk in macosx iphoneos; do
```

File: rust/connlib/clients/apple/build-xcframework-dev.sh
Comment (roop) [inline rust/connlib/clients/apple/build-xcframework-dev.sh:24]:
```suggestion
  -framework ./connlib-iphoneos.xcarchive/Products/Library/Frameworks/connlib.framework \
  -framework ./connlib-macosx.xcarchive/Products/Library/Frameworks/connlib.framework \
```

File: rust/connlib/clients/apple/build-xcframework-dev.sh
Comment (roop) [inline rust/connlib/clients/apple/build-xcframework-dev.sh:29]:
```suggestion
rm -rf ./connlib-iphoneos.xcarchive
rm -rf ./connlib-macosx.xcarchive
```

File: rust/connlib/clients/apple/build-xcframework-dev.sh
Comment (jamilbk) [reply rust/connlib/clients/apple/build-xcframework-dev.sh]: Ah, thanks for catching this @roop; yes on Apple I think we decided to develop for both `iphoneos` and `macosx` targets in parallel due to how much overlap there is. The main difference I believe is only the UI (and the hurdles involved with debugging on the iphonesimulator)

**Review by conectado — DISMISSED** [review]
CI is passing, woo!

Excited to get this merged

**Comment by francesca64** [comment]
@roop good catch! I've added `on_disconnect` to the API, but it currently never gets called by connlib; implementing that will be left for a future PR. Logging will likewise be addressed in another future PR.

**Comment by jamilbk** [comment]
Apologies for introducing the codespell issue. Rebased which should fix it. Will merge when green.

**Review by jamilbk — APPROVED** [review]

**Comment by jamilbk** [comment]
Think it's now waiting on @AndrewDryga for some odd reason. Perhaps because of the rebase.

**Review by AndrewDryga — APPROVED** [review]

**Comment by jamilbk** [comment]
Waiting on @conectado's final approval 😆

**Review by conectado — APPROVED** [review]
:tada:

**Pull request merged**
